### PR TITLE
[CCORE-282] Working around Dalli interaction with Elasticache memcach…

### PR DESCRIPTION
**Pull request by snbloch**
…ed nodes being rebooted

See: https://github.com/petergoldstein/dalli/issues/390

@grosser @bgatelet Please have a look 👀

### CODE REVIEW

**Comment by snbloch** [comment]
# Running:

/Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
.../Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
./Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
./Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
./Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
.../Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
../Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
./Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
./Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
.../Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
.../Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
....................../Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
...................................../Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
./Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
......./Users/steven.bloch/.rbenv/versions/2.6.8/lib/ruby/gems/2.6.0/gems/test_after_commit-0.4.2/lib/test_after_commit.rb:55: warning: instance variable @transaction not initialized
...

Finished in 0.623490s, 142.7449 runs/s, 336.8137 assertions/s.

89 runs, 210 assertions, 0 failures, 0 errors, 0 skips

**Review by cryptomail — APPROVED** [review]
Love it!
Hmm....any tests though?

**Review by cryptomail — CHANGES_REQUESTED** [review]
I will try to make so tests that exercise this

File: test/find_some_test.rb [hunk 1]
```diff
@@ -66,6 +66,37 @@
     assert_equal [@post1, @post2].map(&:id).sort, found_posts.map(&:id).sort
   end
 
+  describe "tests dalli_allow_true_class_response" do
```
Comment (grosser) [inline test/find_some_test.rb:69]:
```suggestion
  describe "with dalli_allow_true_class_response" do
```

File: test/find_some_test.rb [hunk 2]
```diff
@@ -66,6 +66,37 @@
...
+    before do
+      Kasket::CONFIGURATION[:dalli_allow_true_class_response] = true # default value
+    end
+
+    after do
+      Kasket::CONFIGURATION[:dalli_allow_true_class_response] = true # default value
+    end
+
+    it "Does not raise error when dalli_allow_true_class_response set to false explicitly" do
```
Comment (grosser) [inline test/find_some_test.rb:78]:
```suggestion
    it "does not raise error when set to false" do
```

File: lib/kasket.rb [hunk 3]
```diff
@@ -29,6 +30,7 @@ def setup(options = {})
     CONFIGURATION[:max_collection_size] = options[:max_collection_size] if options[:max_collection_size]
     CONFIGURATION[:write_through]       = options[:write_through]       if options[:write_through]
     CONFIGURATION[:default_expires_in]  = options[:default_expires_in]  if options[:default_expires_in]
+    CONFIGURATION[:dalli_allow_true_class_response] = !!options[:dalli_allow_true_class_response] if options.key?(:dalli_allow_true_class_response)
```
Comment (grosser) [inline lib/kasket.rb:33]: isn't that the same ?
```suggestion
    CONFIGURATION[:dalli_allow_true_class_response] = !!options[:dalli_allow_true_class_response]
```

**Review by grosser — COMMENTED** [review]
dalli_allow_true_class_response sounds more like it would pass true up
maybe should be `dalli_fallback_on_error` or something like that

also sounds like true would never be a valid response since we cache AR objects, so can't it always do the fallback without needing extra config ?

**Comment by cryptomail** [comment]
@grosser we were worried about existing behavior and safety.
If that is NOT a concern, will refactor, will rename, and make this default non switched behavior.

**Comment by grosser** [comment]
better have this as opt-in first to find edge-cases and then promote to default
I'd still do the rename so it's more obvious what is happening under the hood
... if you check the code on what goes into/comes out of Kasket.cache.read and it is only AR objects then I'd say it's safe to be turned into default right away

**Comment by grosser** [comment]
... the code would have always blown up with true, so I think this is safe to add without a guard

**Comment by grosser** [comment]
https://api.rubyonrails.org/v6.1.4/classes/ActiveRecord/Querying.html#method-i-find_by_sql always returns an array as per the docs ... so we'd never store a true, so we can never pull out a true unless it's an error 🤷

**Comment by cryptomail** [comment]
@grosser Got it, that's pretty reassuring and we kind of thought that, but wanted to ensure all AR versions had that return value.  I'm fairly certain if AR at one time had a differing return value (not just array type), it must have been ANCIENT history and a fair risk for consumers of this library if they're going back **that** far :)
So, I removed the guard, and made a poisoned cache test.

File: lib/kasket/read_mixin.rb [hunk 4]
```diff
@@ -33,7 +33,16 @@ def find_by_sql_with_kasket(*args)
...
-            if value.is_a?(Array)
+            # Identified a specific edge case where memcached server returns 0x00 binary protocol response with no data
+            # when the node is being rebooted which causes the Dalli memcached client to return a TrueClass object instead of nil
+            # see: https://github.com/petergoldstein/dalli/blob/31dabf19d3dd94b348a00a59fe5a7b8fa80ce3ad/lib/dalli/server.rb#L520
+            # and: https://github.com/petergoldstein/dalli/issues/390
+            #
+            # The code in this first condition of TrueClass === true  will
+            # skip the kasket cache for these specific objects and go directly to SQL for retrieval.
+            if value.is_a?(TrueClass)
+              find_by_sql_without_kasket(*args)
```
Comment (mervync) [inline lib/kasket/read_mixin.rb:44]: (nit) maybe log a warning?

**Review by mervync — APPROVED** [review]
Ugh, it seems inappropriate to workaround this problem at this level. I'd expect

**Review by cryptomail — APPROVED** [review]
made tests
